pkg/backup/backupsink/sst_sink_key_writer.go

```diff
@@ -0,0 +1,310 @@
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package backupsink
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/cockroachdb/cockroach/pkg/backup/backuppb"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/storage"
+	"github.com/cockroachdb/cockroach/pkg/util/admission"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+)
+
+type SSTSinkKeyWriter struct {
+	FileSSTSink
+	// prevKey represents the last key written using WriteKey. When writing a new
+	// key, this helps determine if the last key written was mid-row. This resets
+	// on each flush.
+	prevKey roachpb.Key
+}
+
+func MakeSSTSinkKeyWriter(
+	conf SSTSinkConf, dest cloud.ExternalStorage, pacer *admission.Pacer,
+) *SSTSinkKeyWriter {
+	return &SSTSinkKeyWriter{
+		FileSSTSink: *MakeFileSSTSink(conf, dest, pacer),
+	}
+}
+
+// WriteKey writes a single key to the SST file. The key should be the full key.
+// span refers to the span that the key belongs to. start and end are the time
+// bounds of the span being backed up. The writing of each key could
+// potentially split up the span of the previously written key (if both keys
+// were part of the same span and the span had to be split). As a consequence
+// the span that is recorded for the new key is returned, as it may be a subspan
+// of the span passed in for the key.
+//
+// Flush should be called after the last key is written to ensure that the SST
+// is written to the destination.
+func (s *SSTSinkKeyWriter) WriteKey(ctx context.Context, key storage.MVCCKey, value []byte) error {
+	if len(s.flushedFiles) == 0 {
+		return errors.AssertionFailedf(
+			"no BackupManifest_File to write key to, call Reset before WriteKey",
+		)
+	}
+	if err := s.maybeFlushOOOKey(ctx, key.Key); err != nil {
+		return err
+	}
+	s.setMidRowForPrevKey(key.Key)
+	err := s.maybeDoSizeFlushWithNextKey(ctx, key.Key)
+	if err != nil {
+		return err
+	}
+
+	// At this point, because of the Reset invariant, we can make the following
+	// assumptions about the new key:
+	// - The new key comes after the previous key.
+	// - This new key belongs to the span of the last BackupManifest_File in
+	//   s.flushedFiles.
+	if len(s.flushedFiles) == 0 {
+		panic(
+			"unexpected missing BackupManifest_File after flushes when writing key",
+		)
+	}
+	elidedKey, prefix, err := elideMVCCKeyPrefix(key, s.conf.ElideMode)
+	if err != nil {
+		return err
+	}
+	s.elidedPrefix = append(s.elidedPrefix[:0], prefix...)
+	if err := s.sst.PutRawMVCC(elidedKey, value); err != nil {
+		return err
+	}
+
+	// TODO (kev-cao): Should add some method of tracking `Rows` (and potentially
+	// `IndexEntries`).
+	keyAsRowCount := roachpb.RowCount{
+		DataSize: int64(len(key.Key)) + int64(len(value)),
+	}
+	lastFile := &s.flushedFiles[len(s.flushedFiles)-1]
```

Comment:
we have to do file entry flushing before we write the key the sst. Currently, the written key is equal to the _exclusive_ endkey of the prev file entry and the _inclusive_ start key of the next file entry but is written to the prevfile entry. This could lead to corruption.

Reply:
you could fold this into the `maybeDoSizeFlushWithNextKey` to be:
```
lastFileEndKey := &s.flushedFiles[len(s.flushedFiles)-1].Span.EndKey
if shouldDoSizedBaseFlush(){
  s.doFlush()
} else if s.flushedFiles[len(s.flushedFiles)-1].EntryCounts.DataSize >= fileSpanByteLimit {
} else {
 return
}
newSpan := roachpb.Span{
	Key:    nextKey,
	EndKey: *lastFileEndKey,
}
 return s.Reset(ctx, newSpan)
```

Reply:
Oh yeah good catch, I'll write a unit test for this too, should've caught it.

Reply:
As for folding it into it, I think it'd be better just to move the logic into a separate function? While both are size flushes, one just creates a new `BackupManifest_File` where the other opens a new SST, I think it'd be better to keep those two concerns separate.

Reply:
> While both are size flushes, one just creates a new BackupManifest_File where the other opens a new SST, I think it'd be better to keep those two concerns separate.

I prefer the tighter coupling I propose above because I think it is easier to read: it suggests that on a given `WriteKey`, you can either flush  the sst, and if not, you can flush the file entry. If you have 2 separate functions, theres more ambiguity around doing both kinds of flushes. Furthermore, `Reset` already handles both cases: if `s.out` is nil, we simply open a new sst. Happy to chat more about this over a call.